Restructure assistant edits to show all changes in a proposed-change editor

maxbrunsfeld opened the pull request:
This changes the `/workflow` command so that instead of emitting edits in separate steps, the user is presented with a single tab, with an editable diff that they can apply to the buffer.

Todo

* Assistant panel
    * [x] Show a patch title and a list of changed files in a block decoration
    * [x] Don't store resolved patches as state on Context. Resolve on demand.
    * [ ] Better presentation of patches in the panel
    * [ ] Show a spinner while patch is streaming in
* Patches
    * [x] Preserve leading whitespace in new text, auto-indent insertions
    * [x] Ensure patch title is very short, to fit better in tab
    * [x] Improve patch location resolution, prefer skipping whitespace over skipping `}`
    * [x] Ensure patch edits are auto-indented properly
    * [ ] Apply `Update` edits via a diff between the old and new text, to get fine-grained edits.
* Proposed changes editor
    * [x] Show patch title in the tab
    * [x] Add a toolbar with an "Apply all" button
    * [x] Make `open excerpts` open the corresponding location in the base buffer (https://github.com/zed-industries/zed/pull/18591)
    * [x] Add an apply button above every hunk (https://github.com/zed-industries/zed/pull/18592)
    * [x] Expand all diff hunks by default (https://github.com/zed-industries/zed/pull/18598)
    * [x] Fix https://github.com/zed-industries/zed/issues/18589
    * [x] Syntax highlighting doesn't work until the buffer is edited (https://github.com/zed-industries/zed/pull/18648)
    * [x] Disable LSP interaction in Proposed Changes editor (https://github.com/zed-industries/zed/pull/18945)
    * [x] No auto-indent? (https://github.com/zed-industries/zed/pull/18984)
* Prompt
    * [ ] make sure old_text is unique

Release Notes:

- N/A

maxdeviant committed:
Start restructing edits to use `old_text` and `new_text`

Co-authored-by: Antonio <antonio@zed.dev>
Co-authored-by: Max <max@zed.dev>
Co-authored-by: Richard <richard@zed.dev>

maxbrunsfeld committed:
Rename step -> patch

maxdeviant committed:
Merge branch 'main' into single-shot-edits

maxbrunsfeld committed:
Merge branch main into single-shot-edits

maxbrunsfeld committed:
Start work on using ProposedChangesEditor for patches provided by assistant

maxbrunsfeld committed:
Use a single crease per assistant patch

maxbrunsfeld committed:
Add newlines for insert_before, insert_after

maxbrunsfeld committed:
Merge branch 'main' into single-shot-edits

maxbrunsfeld committed:
Add a title to assistant patches, display them in the assistant panel

Co-authored-by: Marshall <marshall@zed.dev>

maan2003 commented on assets/prompts/edit_workflow.hbs:
```diff
@@ -1,85 +1,35 @@
+        - <new_text> (required) - The new text to insert at the given location in the file.
+        - <operation> (required) - The type of change that should happen at this range of the file.
+          Must be one of the following values:
+            - `update`: Rewrites the specified string entirely based on the given description.
```
"specified string" is not well defined before this, llm may get confused.

maxbrunsfeld committed:
Merge branch 'main' into single-shot-edits

maxbrunsfeld committed:
Merge branch 'main' into single-shot-edits

maxbrunsfeld committed:
Merge branch 'main' into single-shot-edits

maxdeviant committed:
Merge branch 'main' into single-shot-edits

maxbrunsfeld committed:
Merge branch 'main' into single-shot-edits

maxbrunsfeld committed:
Merge branch 'main' into single-shot-edits

maxbrunsfeld committed:
Don't trim leading whitespace from new text in patch edits

maxbrunsfeld committed:
Merge branch 'main' into single-shot-edits

iamnbutler committed:
Add diff icon, apply to proposed changes editor

maxbrunsfeld committed:
Update diff editors when patches are updated in context

maxbrunsfeld committed:
Merge branch 'main' into single-shot-edits

maxbrunsfeld committed:
Refine workflow prompt

maxbrunsfeld committed:
Rename workflow module to patch, rename types within

maxbrunsfeld committed:
Represent resolved assistant edits as a struct, not an enum

maxbrunsfeld committed:
Fix compile error when building tests depending on languages crate

maxbrunsfeld committed:
Make edit resolution more testable, add failing test for location resolution issue

maxbrunsfeld committed:
WIP - Make location resolution prefer to change whitespace vs non-whitespace bytes

as-cii committed:
Rework `resolve_location` to work with whitespace deletion/insertion

as-cii committed:
Fix autoindentation test

as-cii committed:
Don't specify `original_indent_columns`

This makes the auto-indent logic work correctly when the assistant
suggests edits.

as-cii committed:
Rework styling of live diffs in assistant panel

as-cii committed:
Hide footer when expanding the raw patch XML

rtfeldman committed:
Add settings for experimental live diffs.

Co-authored-by: Antonio <antonio@zed.dev>
Co-authored-by: Marshall <marshall@zed.dev>

rtfeldman committed:
Remove unused import

rtfeldman committed:
Enable /workflow only when the setting is enabled

rtfeldman committed:
Enable /workflow for staff, or iff setting is enabled

rtfeldman committed:
Merge remote-tracking branch 'origin/main' into single-shot-edits

rtfeldman committed:
Fix conflicting Debug implementations

rtfeldman commented on crates/gpui/src/app/entity_map.rs:
```diff
@@ -576,7 +567,10 @@ pub struct WeakModel<T> {
+        f.debug_struct(&type_name::<Self>())
+            .field("entity_id", &self.any_model.entity_id)
+            .field("entity_type", &type_name::<T>())
+            .finish()
```
FYI @mikayla-maki @maxbrunsfeld I'm assuming this conflicting impl just resulted from both `main` and this branch independently adding a `Debug` impl, and I resolved it by taking elements of both impls that seemed nice to me.

rtfeldman committed:
Fix diff parsing tests

Co-authored-by: Antonio <antonio@zed.dev>
Co-authored-by: Marshall <marshall@zed.dev>
